File: FinalProject/GameLogic/GameState.cs
/*
 * Module: GameState.cs
 * Purpose: Manages the persistent state of the game.
 * Functionality:
 *  - Stores player inventory, actions taken, and visited flags.
 *  - Handles serialization (Save) and deserialization (Load) of game data to JSON.
 *  - Ensures game progress is preserved between sessions.
 */
using System;
using System.Collections.Generic;
using System.Text.Json;
using FinalProject.Classes;

namespace FinalProject.GameLogic
{
    public class GameState
    {
        public int PlayerX { get; set; }
        public int PlayerY { get; set; }

        // Track inventory and actions
        public List<string> Inventory { get; set; } = new();
        public List<string> ActionsTaken { get; set; } = new();

        // Remember selected player
        public Player CurrentPlayer { get; set; }

        // Use this to show start message ONLY when the is starting.
        public bool HasSeenStartMessage { get; set; } = false;

        public int StoredQHealth { get; set; } = -1;

        // Save the game state to a JSON file
        public void Save(string filePath)
        {
            var options = new JsonSerializerOptions { WriteIndented = true };
            string json = JsonSerializer.Serialize(this, options);
            File.WriteAllText(filePath, json);
        }

        // Load the game state from a JSON file
        public static GameState Load(string filePath)
        {
            if (!File.Exists(filePath))
                return new GameState();

            string json = File.ReadAllText(filePath);
            return JsonSerializer.Deserialize<GameState>(json) ?? new GameState();
        }
    }

}

File: js/GameState.js
/**
 * GameState.js — Persistent game state (port of GameLogic/GameState.cs)
 * 
 * Uses localStorage instead of JSON file I/O.
 */
class GameState {
    constructor() {
        this.playerX = 0;
        this.playerY = 0;
        this.inventory = [];
        this.actionsTaken = [];
        this.hasSeenStartMessage = false;
        this.storedQHealth = -1;
        this.playerData = null; // serialized player
    }

    save() {
        const data = {
            playerX: this.playerX,
            playerY: this.playerY,
            inventory: this.inventory,
            actionsTaken: this.actionsTaken,
            hasSeenStartMessage: this.hasSeenStartMessage,
            storedQHealth: this.storedQHealth,
            playerData: this.playerData
        };
        localStorage.setItem('starTrekGameState', JSON.stringify(data));
    }

    static load() {
        const json = localStorage.getItem('starTrekGameState');
        if (!json) return null;

        try {
            const data = JSON.parse(json);
            const state = new GameState();
            state.playerX = data.playerX || 0;
            state.playerY = data.playerY || 0;
            state.inventory = data.inventory || [];
            state.actionsTaken = data.actionsTaken || [];
            state.hasSeenStartMessage = data.hasSeenStartMessage || false;
            state.storedQHealth = data.storedQHealth || -1;
            state.playerData = data.playerData || null;
            return state;
        } catch (e) {
            console.error('Failed to load game state:', e);
            return null;
        }
    }

    static clear() {
        localStorage.removeItem('starTrekGameState');
    }

    static exists() {
        return localStorage.getItem('starTrekGameState') !== null;
    }
}
